### ISSUE TITLE
Upgrading from v4.0.1 to 5.0.0 causes some issues related to computed props

### ISSUE THREAD

**Issue by jmyrland** [ctrlplusb/easy-peasy#633]
Upgrading to the newest release of easy-peasy has broken some of our tests:
```
Test Suites: 31 failed, 34 passed, 65 total
Tests:       145 failed, 246 passed, 391 total
```

Upon further inspection, it seems that there is some issues with some `computed` properties being undefined (when they are not supposted to be).

Here is an excerpt of one of the repeating errors:
```
  TypeError: Cannot read property 'map' of undefined
  
      at actionReducer (src/components/App/Local.model.ts:109:52)  
      at fn (../../../node_modules/easy-peasy/src/create-reducer.js:8:43)
      at simpleProduce (../../../node_modules/easy-peasy/src/lib.js:124:20)
      at runActionReducerAtPath (../../../node_modules/easy-peasy/src/create-reducer.js:8:5)      at reducerForActions (../../../node_modules/easy-peasy/src/create-reducer.js:13:14)   
      at rootReducer (../../../node_modules/easy-peasy/src/create-reducer.js:37:31)
      at next (../../../node_modules/redux/lib/redux.js:219:22)
      at next (../../../node_modules/easy-peasy/src/persistence.js:236:19)
      at next (../../../node_modules/easy-peasy/src/effects.js:6:14)
      at ../../../node_modules/easy-peasy/src/listeners.js:5:20
      at next (../../../node_modules/redux-thunk/lib/index.js:14:16)
      at Object.dispatch (../../../node_modules/easy-peasy/src/computed-properties.js:29:12)      at Object.setGeneration (../../../node_modules/easy-peasy/src/actions.js:12:15
```

Has anything changed in how computed properties work, from v4 to v5?

**Comment by ctrlplusb** [comment]
Thanks for the report! Will look into this tomorrow. Nothing significant changed so likely something silly I did. Apologies

**Comment by ctrlplusb** [comment]
Fixed in 5.0.1 🙏

Had to do with some redux internals updates.

**Comment by jmyrland** [comment]
Thanks for your dedication & support @ctrlplusb, I do appreciate it! 🙏 

I've updated to `5.0.1`:
```
easy-peasy@5.0.1:
  version "5.0.1"
  dependencies:
    "@babel/runtime" "^7.13.7"
    immer "^8.0.1"
    redux "^4.0.5"
    redux-thunk "^2.3.0"
    ts-toolbelt "^9.3.12"
```

but I get the exact same result when running the tests:
```
Test Suites: 31 failed, 34 passed, 65 total
Tests:       145 failed, 246 passed, 391 total
```

At first glance, it seems that it is the same error as first posted:
```
    TypeError: Cannot read property 'map' of undefined

    > 109 |         const firstGeneration = state.computedProp.map(node => node.value);     
          |                                                    ^

      at actionReducer (src/components/App/Some.model.ts:109:52)        
      at fn (../../../node_modules/easy-peasy/src/create-reducer.js:8:43)
      at simpleProduce (../../../node_modules/easy-peasy/src/lib.js:124:20)
      at runActionReducerAtPath (../../../node_modules/easy-peasy/src/create-reducer.js:8:5)      
      at reducerForActions (../../../node_modules/easy-peasy/src/create-reducer.js:13:14)   
      at rootReducer (../../../node_modules/easy-peasy/src/create-reducer.js:37:31)
      at next (../../../node_modules/redux/lib/redux.js:219:22)
      at next (../../../node_modules/easy-peasy/src/persistence.js:236:19)
      at next (../../../node_modules/easy-peasy/src/effects.js:6:14)
      at ../../../node_modules/easy-peasy/src/listeners.js:5:20
      at next (../../../node_modules/redux-thunk/lib/index.js:14:16)
      at Object.dispatch (../../../node_modules/easy-peasy/src/computed-properties.js:29:20)      
      at Object.setGeneration (../../../node_modules/easy-peasy/src/actions.js:12:15)  
  ```

**Comment by jmyrland** [comment]
I've created a sandbox to reproduce this issue: https://codesandbox.io/s/naughty-voice-62zy6?file=/src/App.model.test.tsx

![image](https://user-images.githubusercontent.com/837651/109288807-a60e2500-7825-11eb-88a4-7031f564238e.png)

Disclaimer: It is a contrived example to illustrate the error.

If you switch to `easy-peasy@4.0.1`, the tests are passing:
![image](https://user-images.githubusercontent.com/837651/109288696-81b24880-7825-11eb-86a2-445ad516a2a9.png)

**Comment by ctrlplusb** [comment]
Hi @jmyrland 

I have published a new alpha (5.0.2-alpha.0) - gonna try it against your sandbox. Thanks for that.

**Comment by ctrlplusb** [comment]
<img width="619" alt="Screenshot 2021-02-26 at 13 10 07" src="https://user-images.githubusercontent.com/12164768/109304296-f68c7f00-7833-11eb-8004-cb9ac98c56ab.png">

😎

**Comment by ctrlplusb** [comment]
Can you please try it in your project and let me know. I will publish a new version if all good. 🤙

**Comment by jmyrland** [comment]
Unfortunately, `5.0.2-alpha.0` has even generated more errors:
```
Test Suites: 43 failed, 22 passed, 65 total
Tests:       190 failed, 225 passed, 415 total
```

I get repeating errors of the following kind:
```
    TypeError: Cannot read property 'getState' of undefined

    > 47 |   const state = useStoreState((store: ISomeModel) => store.someState);
         |                 ^

      at useStoreState (../../shared/node_modules/easy-peasy/src/hooks.js:38:43)
      at SectionForm (../../shared/components/SectionForm/SectionForm.tsx:47:22)
      at renderWithHooks (../../../node_modules/react-dom/cjs/react-dom.development.js:14985:18)
      at mountIndeterminateComponent (../../../node_modules/react-dom/cjs/react-dom.development.js:17811:13)
      at beginWork (../../../node_modules/react-dom/cjs/react-dom.development.js:19049:16)
      at HTMLUnknownElement.callCallback (../../../node_modules/react-dom/cjs/react-dom.development.js:3945:14)
      at HTMLUnknownElement.callTheUserObjectsOperation (../../../node_modules/jsdom/lib/jsdom/living/generated/EventListener.js:26:30)
      at innerInvokeEventListeners (../../../node_modules/jsdom/lib/jsdom/living/events/EventTarget-impl.js:318:25)
      at invokeEventListeners (../../../node_modules/jsdom/lib/jsdom/living/events/EventTarget-impl.js:274:3)
      at HTMLUnknownElementImpl._dispatch (../../../node_modules/jsdom/lib/jsdom/living/events/EventTarget-impl.js:221:9)
      at HTMLUnknownElementImpl.dispatchEvent (../../../node_modules/jsdom/lib/jsdom/living/events/EventTarget-impl.js:94:17)
      at HTMLUnknownElement.dispatchEvent (../../../node_modules/jsdom/lib/jsdom/living/generated/EventTarget.js:231:34)
      at Object.invokeGuardedCallbackDev (../../../node_modules/react-dom/cjs/react-dom.development.js:3994:16)
      at invokeGuardedCallback (../../../node_modules/react-dom/cjs/react-dom.development.js:4056:31)
      at beginWork$1 (../../../node_modules/react-dom/cjs/react-dom.development.js:23964:7)
      at performUnitOfWork (../../../node_modules/react-dom/cjs/react-dom.development.js:22779:12)
      at workLoopSync (../../../node_modules/react-dom/cjs/react-dom.development.js:22707:5)
      at renderRootSync (../../../node_modules/react-dom/cjs/react-dom.development.js:22670:7)
      at performSyncWorkOnRoot (../../../node_modules/react-dom/cjs/react-dom.development.js:22293:18)
      at scheduleUpdateOnFiber (../../../node_modules/react-dom/cjs/react-dom.development.js:21881:7)
      at updateContainer (../../../node_modules/react-dom/cjs/react-dom.development.js:25482:3)
      at ../../../node_modules/react-dom/cjs/react-dom.development.js:26021:7
      at unbatchedUpdates (../../../node_modules/react-dom/cjs/react-dom.development.js:22431:12)
      at legacyRenderSubtreeIntoContainer (../../../node_modules/react-dom/cjs/react-dom.development.js:26020:5)
      at Object.render (../../../node_modules/react-dom/cjs/react-dom.development.js:26103:10)
      at ../../../node_modules/@testing-library/react/dist/pure.js:99:25
      at batchedUpdates$1 (../../../node_modules/react-dom/cjs/react-dom.development.js:22380:12)
      at act (../../../node_modules/react-dom/cjs/react-dom-test-utils.development.js:1042:14)
      at render (../../../node_modules/@testing-library/react/dist/pure.js:95:26)
```

All of the errors seems to be related to the `useStoreActions` and `useStoreState` from `easy-peasy`.

I'll try to update the snapshot to get a reproducible result 👍

**Comment by ctrlplusb** [comment]
Wow, a bit scary. I have a fairly robust set of unit tests that are all passing. I'll need to develop some more cases for this issue for sure.

**Comment by ctrlplusb** [comment]
That example error you have just shown me is typical of when a StoreProvider hasn't been set up correctly. i.e. The store instance isn't available. 

It could also mean you have two versions of Easy Peasy being referenced. I would also try a node_modules nuke and double check your dependencies. 🙏

**Comment by ctrlplusb** [comment]
> const state = useStoreState((store: ISomeModel) => store.someState);

Make sure to see the `createTypedHooks` helper to avoid the model passing. :)

**Comment by jmyrland** [comment]
You are correct! We have a neat little mono-repo, and I had missed an easy-peasy update, resulting in 2 versions 🤯

The results is much better! However, there is still a few failing tests:
```
Test Suites: 2 failed, 63 passed, 65 total
Tests:       5 failed, 426 passed, 431 total
```

Something is `undefined` when it is not supposted to be - but it is not quite clear what is causing the issue, so I'll have to investigate it a little  further.

I don't know if you can see any issues from the stack trace?

```
      at someFunction (src/components/App/SomeModel.model.ts:132:15)
      at filter (src/components/App/SomeModel.model.ts:93:69)
          at Proxy.filter (<anonymous>)
      at actionReducer (src/components/App/SomeModel.model.ts:93:56)
      at fn (../../../node_modules/easy-peasy/src/create-reducer.js:8:43)
      at simpleProduce (../../../node_modules/easy-peasy/src/lib.js:124:20)
      at runActionReducerAtPath (../../../node_modules/easy-peasy/src/create-reducer.js:8:5)
      at reducerForActions (../../../node_modules/easy-peasy/src/create-reducer.js:13:14)
      at rootReducer (../../../node_modules/easy-peasy/src/create-reducer.js:37:31)
      at next (../../../node_modules/redux/lib/redux.js:219:22)
      at next (../../../node_modules/easy-peasy/src/persistence.js:236:19)
      at next (../../../node_modules/easy-peasy/src/effects.js:6:14)
      at ../../../node_modules/easy-peasy/src/listeners.js:5:20
      at next (../../../node_modules/redux-thunk/lib/index.js:14:16)
      at Object.dispatch (../../../node_modules/easy-peasy/src/computed-properties.js:32:20)
      at Object.setGeneration (../../../node_modules/easy-peasy/src/actions.js:12:15)
```

**Comment by ctrlplusb** [comment]
Making progress!

Can you please try `easy-peasy@5.0.2-alpha.1` 🤞

**Comment by jmyrland** [comment]
Unfortunately no changes from `easy-peasy@5.0.2-alpha.0` to `easy-peasy@5.0.2-alpha.1`; still 5 failing tests. 

Trying to gather some more info based on the test, but I'll have to debug a little further.

**Comment by jmyrland** [comment]
So, I've got an action that uses `state.values`, a computed array of "stuff". When I pass the `state.values` into a function, that `reduces` the array - at this point the test is crashing, because it is stating that `state.values` is `undefined`, hence the error message:
> TypeError: Cannot read property 'reduce' of undefined

Now, the fun part - if i `console.log(state.values)` at the start of the action, the value **is logged _and the test is suddenly passing_**! 🤯 🤯 🤯 
If I remove the `console.log` again, the test fails - because `state.values` is undefined.

`state.values` is [schrödingers cat](https://en.wikipedia.org/wiki/Schr%C3%B6dinger%27s_cat) 🤣 

Do you make any sense of this? Should I try to create a reproducible example?

**Comment by ctrlplusb** [comment]
> state.values is schrödingers cat 🤣

Hahahaha, that is classic. Also super frustrating.

That is quite a detailed breakdown though. Thank you! I will try use this information and attempt to analyse some potential code paths that could cause this behaviour.

**Comment by ctrlplusb** [comment]
Please try latest. 🤞

**Comment by jmyrland** [comment]
Tested `easy-peasy@5.0.2` with the following results:
```
Test Suites: 9 failed, 49 passed, 58 total
Tests:       14 failed, 403 passed, 417 total
```

I'm not quite sure what is happening now; we got some snapshot tests to verify that we get the correct results from computed props in the store. Some of these are now missing a property:
```test
    expect(received).toEqual(expected) // deep equality

    - Expected  - 1
    + Received  + 0

    @@ -2,7 +2,6 @@
        "property1": "some value",
    -   "property2": "some other value",
      }
```

This one is also a headscratcer:
```
    expect(received).toMatchSnapshot()

    - Snapshot  - 0
    + Received  + 8

    @@ -1,6 +1,14 @@
      Object {
    +   "1": Object {...},
    +   "2": Object {...},
        "3": Object {...},
        "4": Object {

      93 | 
      94 |         expect(Object.keys(dictionary)).not.toContain(['1', '2']); // 👈 🤯
    > 95 |         expect(dictionary).toMatchSnapshot();
         |                             ^
      96 |       });
      97 |     });
      98 | 
```

As you can see, on line 94, we have an explicit check that `dicitonary` should not contain 1 or 2, which passes. However, on the next line, the snapshot is containing these values 🤔

It also seems that, after "seeding" the store, the tests does not retain the set values?
```diff
    expect(received).toMatchSnapshot()

    - Snapshot  - 1
    + Received  + 1

    @@ -36,11 +36,11 @@
                      <h2
                        class="mb-0 font-bold text-md"
                      >
    -                   Rocky Balboa
    +                   undefined undefined
                      </h2>

      37 | 
      38 |     const { container } = render(withRouter(withStore(<Component />, store)));
    > 39 |     expect(container).toMatchSnapshot();
         |                       ^
      40 |   });
      41 | });
      42 | 
```

Also in, some cases,  the expected values are completely wrong:
```
    expect(received).toMatchSnapshot()

    - Snapshot  - 2
    + Received  + 2

    @@ -10,13 +10,13 @@
          <p>
    -       Philip J.
    +       New
             
    -       Fry
    +       Guy
          </p>

      55 |       });
      56 |       const { container } = render(withStore(<AnotherComponent />, store));
    > 57 |       expect(container).toMatchSnapshot();
         |                         ^
      58 |     });
```

In general it seems that the tests are failing, because the expected store values is not "as expected". The wierd thing is that this is only happening in 14 (out of 417) tests, working on different parts of the store - and I cannot see any commonalities between them.

**Comment by ctrlplusb** [comment]
Sorry for the back and forth. This is real shrödinger stuff here!

Can you please try `easy-peasy@5.0.3`

**Comment by jmyrland** [comment]
> Sorry for the back and forth. 

First off, I don't want you to feel that this is a critical issue for us; we are in no rush - so please take the time you need and don't feel "forced" to fix this.

Because we have such extensive testing related to stores and models, I think it's a good "check" to have for new releases. In other words, I want to contribute to ensure that new releases is "backwards compatible" (if not stated otherwise).

TLDR; I really don't mind the back and forth. Take your time, we are in no rush to fix this, I just want to contribute 🙂

> This is real shrödinger stuff here!

Haha! 😆

> Can you please try `easy-peasy@5.0.3`
 
I'll check it out and report back soon 👌

**Comment by jmyrland** [comment]
We got some progress 👏 

```
Test Suites: 1 failed, 57 passed, 58 total
Tests:       1 failed, 416 passed, 417 total
```

Just a single test to go 😁 

```
    expect(received).toHaveProperty(path)

    Expected path: "x"
    Received path: []

    Received value: {"1": {...}, "1.2": {...}, "2": {...}} 

      233 |         } = store.getState();
      234 | 
    > 235 |         expect(map).toHaveProperty('x');
          |                             ^
```

I've reconstructed the model & related test in the [sandbox](https://codesandbox.io/s/naughty-voice-62zy6?file=/src/Tree.model.test.ts), and it generates the same error when using `easy-peasy@5.0.2` - but it passes with `easy-peasy@5.0.3` 🤔 However, I still get the error locally.

The only difference in functionality between the sandbox model and the local (failing) modal, is that the local modal has some dependencies of a computed prop from another store.

I'll tinker a little more with it, and see if I manage to reproduce the issue in the sandbox.

**Comment by jmyrland** [comment]
Across our mono repo, we have a total of 884 unit tests - most of them touches `easy-peasy` in some shape or form. I've run the test for the entire repo, and **_only a single test is failing_**.

I've been scratching my head for a while, to figure out why the single test is failing. I have tried reproducing the exact conditions in the sandbox, but the tests are still passing there.

I just tried debugging the failing test - and of course, we got a schrödingers situation here. When I debug - the tests are of course passing 👯‍♂️

I've tried to insert a delay between the action and `getState` - but that had no effect on the result.

If you got any pointers on what I could try next, I'm all ears!

**Comment by jmyrland** [comment]
I'm stuck on this one. For some reason, I'm unable to reproduce the issue we're having. In [this sandbox](https://codesandbox.io/s/modest-wind-k0ym8?file=/src/App.model.test.tsx), everyting is working perfectly - but I was expecting the tests to fail, due to "stale" computed values.

In my actual failing test, if I just insert a targeted `getState()` on the computed property, the issue is resolved. I am tempted to just leave it this way 🤔

Would you consider it safe to upgrade? Could this potentially cause some issues in a real scenario?
I am not (yet) able to produce any issues or errors when running the app.

**Comment by jmyrland** [comment]
We have upgraded to 5.0.3 and tested our entire solution, and we could not produce any issues. 
Marking this as closed 👍
